sdk/storage/storage-file/src/DirectoryClient.ts

```diff
@@ -202,6 +202,59 @@ export class DirectoryClient extends StorageClient {
     });
   }
 
+  /**
+   * Iterates over containers under the specified account.
+   *
+   * @param {DirectoryListFilesAndDirectoriesSegmentOptions} [options={}] Options to list files and directories(optional)
+   * @returns {AsyncIterableIterator<Models.FileItem | Models.DirectoryItem>}
+   * @memberof DirectoryClient
+   *
+   * @example
+   * let i = 1;
+   * for await (const item of directoryClient.listFilesAndDirectories()) {
+   *   console.log(`${i}: ${item.name}`);
+   *   i++;
+   * }
+   *
+   * @example
+   * let iter1 = directoryClient.listFilesAndDirectories();
+   * let i = 1;
+   * for await (const item of iter1) {
+   *   console.log(`${i}: ${item.name}`);
+   *   i++;
+   * }
+   *
+   * @example
+   * let iter2 = await directoryClient.listFilesAndDirectories();
+   * i = 1;
+   * let item = await iter2.next();
+   * do {
+   *   console.log(`${i++}: ${item.value.name}`);
+   *   item = await iter2.next();
+   * } while (item.value);
+   *
+   */
+  public async *listFilesAndDirectories(
+    options: DirectoryListFilesAndDirectoriesSegmentOptions = {}
+  ): AsyncIterableIterator<Models.FileItem | Models.DirectoryItem> {
+    let marker = undefined;
+    const directoryClient = this;
+    const aborter = !options.abortSignal ? Aborter.none : options.abortSignal;
+    let listFilesAndDirectoriesResponse;
+    do {
+      listFilesAndDirectoriesResponse = await directoryClient.listFilesAndDirectoriesSegment(
+        marker,
+        {
+          ...options,
+          abortSignal: aborter
+        }
+      );
+      marker = listFilesAndDirectoriesResponse.nextMarker;
+      yield* listFilesAndDirectoriesResponse.segment.fileItems;
```

Comment:
I was thinking about adding new classes for the results instead of using the interfaces `DirectoryItem` and `FileItem`, so we can use `instanceof` to check the types of each item from the iterator.

```typescript
for await (const item of directoryClient.listFilesAndDirectories()) {
  if (item instanceof AzureFileItem) {
    console.log(`\tFile: ${item.name}`);
    console.log(`\tFile Properties: ${item.properties}`);
  } else {
    console.log(`\tDirectory: ${item.name}`);
  }
}
```

@bterlson thoughts?

Reply:
It feels important to know what type of thing is getting returned. However, classes may have a lot of extra overhead. Another option is to add a field like `kind` that is either "directory" or "file". TypeScript will type things appropriately inside `if (item.kind === 'directory')` blocks.

Reply:
something like below?

```typescript
alias type ItemKind = "directory" | "file";

interface AzureDirectoryItem {
  kind: ItemKind;
  name: string;
}

interface AzureFileItem {
  kind: ItemKind;
  name: string;
  properties: FileProperty;
}

yield { kind = "directory", name = "dir 1" };

yield { kind = "file", name = "file name 1", properties = p };

```

Reply:
Close. The following definitions of Directory and File are more accurate and will ensure TypeScript's control flow narrowing works properly:

```ts
interface AzureDirectoryItem {
  kind: "directory";
  name: string;
}

interface AzureFileItem {
  kind: "file";
  name: string;
  properties: FileProperty;
}
```

Reply:
Are these definitions only available for `listFilesAndDirectories`?

Reply:
> Are these definitions only available for `listFilesAndDirectories`?

Good point. The other place whether we returns directory/file items is `listFilesAndDirectoriesSegment()`.  Yes it feels a bit inconsistent that here we have an extra field `kind`. On the other hand I don't like transforming the segment response to add `kind` for the items...

Reply:
Will there be intelligence for potential values "directory" or "file", in VSC when accessing `item.kind`?

```javascript
for await (const item of directoryClient.listFilesAndDirectories()) {
  if (item.kind ===
}
```

Reply:
> Will there be intelligence for potential values "directory" or "file",

The tooltip will show them, but there is no completion for them. I think it's good enough as other values will cause compile errors.

Reply:
I get intellisense (completions with both options listed) in VSC. @jeremymeng does nothing happen for you if you hit ctrl+space in that position?

I also don't like modifying the return values. A possibility that is only slightly ridiculous is to use Object.setPrototypeOf and put `kind` on the prototype... It could even be a getter that returns something like `return this.hasOwnProperty('properties') ? 'file' : 'directory'. I don't like this either.

Reply:
> does nothing happen for you if you hit ctrl+space in that position?

just checked again, yes I have "d" and "f" in the completion list, but they appears at the beginning fo the list, while the preselected item is at the middle part so I missed it.

Reply:
So we have some options

1. Don't return new interfaces with extra `kind` field.  Users will have to check the existence of `properties: FileProperty;` field by themselves to tell them apart.

2. Add new interfaces with extra `kind` field.  Users will have to deal with two different set of interfaces: `DirectoryItem` and `FileItem` v.s. `AzureDirectoryItem` and `AzureFileItem` where the latter two has `kind` field. It would be confusing when users see `kind` in one result but not in the other.

3. "use Object.setPrototypeOf and put kind on the prototype... It could even be a getter that returns something like `return this.hasOwnProperty('properties') ? 'file' : 'directory'."  where would we add this?  Right now `DirectoryItem` and `FileItem` are generated by autorest.

4. add new interfaces and transforming the result from `listXxxxSegment` to use new interfaces.

Maybe option 1) for this preview?

Reply:
![image](https://user-images.githubusercontent.com/10452642/58670501-2a9bde80-82f4-11e9-8d6b-618d74cffd27.png)

FileProperty has only one attribute called `contentLength`.

Since, we will be talking about Files and Directories together forever, should we consider merging FileItem and DirectoryItem into one ?
   - We can call the new interface to be `FileOrDirectoryItem` or simply `Entity` [- can correspond to File or Directory]
   - We can add a new attribute called `kind` to that `Entity`.
   - `FilesAndDirectoriesListSegment` will then have `entities` which would point just to `Entity[]` instead of directoryItems, fileItems
   -  Maybe remove `properties` completely and make `contentLength` an optional attribute of `Entity`.


Any comments ?

Reply:
Note that these are generated code that are based on the REST APIs, so it might be more involved for changing the code gen.

Reply:
option 5. use intersection type so the users are clear that we are adding `kind` to the auto-generated items.

```typescript
type AzureDirectoryItem = { kind: "directory"; } & Models.DirectoryItem;
type AzureFileItem = { kind: "directory"; } & Models.FileItem;
```

Reply:
This seems better.

Reply:
Option 5 seems better. `Models.DirectoryItem` is aligned with swagger definitions, and properties of it can be updated anytime when storage changes APIs.

Reply:
Let's go with Option 5.

Reply:
+1 for option 5 (but with the proper kind values ;))

Reply:
oh yeah copy n' paste error...